### DISCUSSION TITLE
Mrgsolve Error in Function

### DISCUSSION THREAD

**Discussion by Kunle-Rhodes** [metrumresearchgroup/mrgsolve#1091] [General]
Hello,
I was trying to do simulations from a two-compartment pop PK model(where I edited the model specification code) and had the following error when I tried to run the simulation with mrgsim command:

**_Error in (function (cond)  : 
  error in evaluating the argument 'x' in selecting a method for function 'plot': pred_VC has a 0 or negative  value._**

It's been difficult to trace where I messed things up. 

Thank you in advance.

**Comment by kylebaron** [comment]
Hi @Kunle-Rhodes - 
Could you please share the R code and model code for this and we can try to help you.
Kyle

**Reply by Kunle-Rhodes** [reply]
$PROB
# Example population PK model

$PARAM 'TVCL'= 87.9, 'TVV1' = 20.5, 'TVV2'=200, 'TVQ'=87.3

$PARAM @covariates
WT = 70

$PKMODEL cmt="CENT PERIPH",depot=FALSE

$SET end=240, delta=0.033

$MAIN
double CL = exp(log(CL) + 0.75*log(WT/70) + ECL);
double V1  = exp(log(V1)  +      log(WT/70) + EV1 );
double Q = exp(log(Q) + 0.75*log(WT/70) + EQ);
double V2  = exp(log(V2)  +      log(WT/70) + EV2 );


$OMEGA @labels ECL EV1 EQ EV2 
0.27 0.96 0.37 0.12 

$SIGMA 0

$TABLE
capture IPRED = CENT/V1;
capture DV = IPRED*exp(EPS(1));

$CAPTURE CL Q V1 V2

**Reply by Kunle-Rhodes** [reply]
R code:
library(mrgsolve)
library(dplyr)
library(xlsx)
library(WriteXLS)
library(ubiquity)

idata<-expand.idata(WT=runif(200,55,100), amt=c(75,225))%>% mutate(dose=amt)
write.xlsx(idata,'numsim1.xlsx', colNames = TRUE)

mod1 <- mread('popex', modlib())

mrgsolve:::details(mod1)
param(mod1)
writeLines(con='poptezn4.cpp',as.list(mod1)$code)

mod3 <- mread('poptezn4')
see(mod3)
param(mod3)
OUTMOD3<- 
  mod3%>%
  ev(amt=75)%>%
  idata_set(idata)%>%
  obsonly()%>%
  #zero_re()%>%
  carry_out(dose)%>%
  mrgsim(start=0, end=24, delta=0.033)%>%
  plot(DV~time)

**Reply by kylebaron** [reply]
I wasn't able to reproduce the error you referenced but did make one change to the code: 

- Don't quote the parameter names in the `$PARAM` block
  - Use this: `TVV1 = 20.5`
  - Not this: `'TVV1' = 20.5`



``` r
code <- '
$PROB

Example population PK model

$PARAM TVCL= 87.9, TVV1 = 20.5, TVV2=200, TVQ=87.3

$PARAM @covariates
WT = 70

$PKMODEL cmt="CENT PERIPH",depot=FALSE

$SET end=240, delta=0.033

$MAIN
double CL = exp(log(TVCL) + 0.75*log(WT/70) + ECL);
double V1 = exp(log(TVV1) + log(WT/70) + EV1 );
double Q = exp(log(TVQ) + 0.75*log(WT/70) + EQ);
double V2 = exp(log(TVV2) + log(WT/70) + EV2 );

$OMEGA @labels ECL EV1 EQ EV2
0.27 0.96 0.37 0.12

$SIGMA 0

$TABLE
capture IPRED = CENT/V1;
capture DV = IPRED*exp(EPS(1));

$CAPTURE CL Q V1 V2
'

mod <- mcode("foo", code)
#> Error in mcode("foo", code): could not find function "mcode"
mrgsim(mod, ev(amt = 100)) %>% plot("CENT")
#> Error in mrgsim(mod, ev(amt = 100)) %>% plot("CENT"): could not find function "%>%"



library(mrgsolve)
#> 
#> Attaching package: 'mrgsolve'
#> The following object is masked from 'package:stats':
#> 
#>     filter
library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union

idata<-expand.idata(WT=runif(200,55,100), amt=c(75,225))%>% mutate(dose=amt)

mod1 <- mread('popex', modlib())
#> Building popex ...
#> done.

mrgsolve:::details(mod1)
#>    block    name descr units options value
#> 1  PARAM    TVKA     .     .       .   0.5
#> 2  PARAM    TVCL     .     .       .   1.0
#> 3  PARAM     TVV     .     .       .  24.0
#> 4  PARAM      WT     .     .       .  70.0
#> 5    CMT     GUT     .     .       .   0.0
#> 6    CMT    CENT     .     .       .   0.0
#> 7  OMEGA     ECL     .     .       .    NA
#> 8  OMEGA  ECL-EV     .     .       .    NA
#> 9  OMEGA      EV     .     .       .    NA
#> 10 OMEGA ECL-EKA     .     .       .    NA
#> 11 OMEGA  EV-EKA     .     .       .    NA
#> 12 OMEGA     EKA     .     .       .    NA
param(mod1)
#> 
#>  Model parameters (N=4):
#>  name value . name value
#>  TVCL 1     | TVV  24   
#>  TVKA 0.5   | WT   70
writeLines(con='poptezn4.cpp',as.list(mod1)$code)

mod3 <- mread('poptezn4')
#> Building poptezn4 ... done.
see(mod3)
#> 
#> Model file:  poptezn4.cpp 
#> $PROB
#> # Example population PK model
#> 
#> $PARAM TVKA = 0.5, TVCL = 1, TVV = 24
#> 
#> $PARAM @covariates
#> WT = 70
#> 
#> $PKMODEL cmt="GUT CENT", depot=TRUE
#> 
#> $SET end=240, delta=0.5
#> 
#> $MAIN
#> double CL = exp(log(TVCL) + 0.75*log(WT/70) + ECL);
#> double V  = exp(log(TVV)  +      log(WT/70) + EV );
#> double KA = exp(log(TVKA)                   + EKA);
#> 
#> $OMEGA @labels ECL EV EKA
#> 0.3 0.1 0.5
#> 
#> $SIGMA 0
#> 
#> $TABLE
#> capture IPRED = CENT/V;
#> capture DV = IPRED*exp(EPS(1));
#> 
#> $CAPTURE CL V ECL
#> 
#> 
param(mod3)
#> 
#>  Model parameters (N=4):
#>  name value . name value
#>  TVCL 1     | TVV  24   
#>  TVKA 0.5   | WT   70

  mod3%>%
  ev(amt=75)%>%
  idata_set(idata)%>%
  obsonly()%>%
  #zero_re()%>%
  carry_out(dose)%>%
  mrgsim(start=0, end=24, delta=0.033)%>%
  plot(DV~time)
```

![](https://i.imgur.com/DOBg2Tg.png)

<sup>Created on 2023-06-27 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Reply by Kunle-Rhodes** [reply]
Thank you so much. I have adjusted the parameter names without quotes, and the error has been resolved.

**Reply by Kunle-Rhodes** [reply]
I have encountered the same problem while trying to simulate with mrgsolve. 

The error message reads **Error: pred_VC has a 0 or negative value.**

 It worked the last time because running the entire line of code changes the modified model to the default one-compartment pop-PK model [not sure why this happens]. When the parameters are viewed, it becomes clear that the model used is a one-compartment model vs. a two-compartment model in the modified model code.

So, there are two main questions:
1. What does pred_VC mean in this case?
2. Why does the modified model code(in the .cpp file) change every time I rerun the entire code?

Thanks in advance
(PS: I am using the exact code attached previously.)

**Reply by kylebaron** [reply]
Hi @Kunle-Rhodes  - 

- `pred_VC` is just an internal variable that gets used for the calculation; it just means volume isn't set correctly
- I don't know which code in the cpp file is getting changed


Can you post a [reproducible example](https://reprex.tidyverse.org/) for the code you are running to get the error? This would help me understand what needs to change to get you running again.


When I run the code you posted, I get this (it works the same as it did before):


``` r
library(mrgsolve)
#> 
#> Attaching package: 'mrgsolve'
#> The following object is masked from 'package:stats':
#> 
#>     filter
library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union

code <- '
$PROB

Example population PK model

$PARAM TVCL= 87.9, TVV1 = 20.5, TVV2=200, TVQ=87.3

$PARAM @covariates
WT = 70

$PKMODEL cmt="CENT PERIPH",depot=FALSE

$SET end=240, delta=0.033

$MAIN
double CL = exp(log(TVCL) + 0.75*log(WT/70) + ECL);
double V1 = exp(log(TVV1) + log(WT/70) + EV1 );
double Q = exp(log(TVQ) + 0.75*log(WT/70) + EQ);
double V2 = exp(log(TVV2) + log(WT/70) + EV2 );

$OMEGA @labels ECL EV1 EQ EV2
0.27 0.96 0.37 0.12

$SIGMA 0

$TABLE
capture IPRED = CENT/V1;
capture DV = IPRED*exp(EPS(1));

$CAPTURE CL Q V1 V2
'

mod <- mcode("foo", code)
#> Building foo ...
#> done.
mrgsim(mod, ev(amt = 100)) %>% plot("CENT")
```

![](https://i.imgur.com/VYtqjQ1.png)

``` r


idata<-expand.idata(WT=runif(200,55,100), amt=c(75,225))%>% mutate(dose=amt)

mod1 <- mread('popex', modlib())
#> Building popex ... done.

mrgsolve:::details(mod1)
#>    block    name descr units options value
#> 1  PARAM    TVKA     .     .       .   0.5
#> 2  PARAM    TVCL     .     .       .   1.0
#> 3  PARAM     TVV     .     .       .  24.0
#> 4  PARAM      WT     .     .       .  70.0
#> 5    CMT     GUT     .     .       .   0.0
#> 6    CMT    CENT     .     .       .   0.0
#> 7  OMEGA     ECL     .     .       .    NA
#> 8  OMEGA  ECL-EV     .     .       .    NA
#> 9  OMEGA      EV     .     .       .    NA
#> 10 OMEGA ECL-EKA     .     .       .    NA
#> 11 OMEGA  EV-EKA     .     .       .    NA
#> 12 OMEGA     EKA     .     .       .    NA
param(mod1)
#> 
#>  Model parameters (N=4):
#>  name value . name value
#>  TVCL 1     | TVV  24   
#>  TVKA 0.5   | WT   70
writeLines(con='poptezn4.cpp',as.list(mod1)$code)

mod3 <- mread('poptezn4')
#> Building poptezn4 ... done.

see(mod3)
#> 
#> Model file:  poptezn4.cpp 
#> $PROB
#> # Example population PK model
#> 
#> $PARAM TVKA = 0.5, TVCL = 1, TVV = 24
#> 
#> $PARAM @covariates
#> WT = 70
#> 
#> $PKMODEL cmt="GUT CENT", depot=TRUE
#> 
#> $SET end=240, delta=0.5
#> 
#> $MAIN
#> double CL = exp(log(TVCL) + 0.75*log(WT/70) + ECL);
#> double V  = exp(log(TVV)  +      log(WT/70) + EV );
#> double KA = exp(log(TVKA)                   + EKA);
#> 
#> $OMEGA @labels ECL EV EKA
#> 0.3 0.1 0.5
#> 
#> $SIGMA 0
#> 
#> $TABLE
#> capture IPRED = CENT/V;
#> capture DV = IPRED*exp(EPS(1));
#> 
#> $CAPTURE CL V ECL
#> 
#> 
param(mod3)
#> 
#>  Model parameters (N=4):
#>  name value . name value
#>  TVCL 1     | TVV  24   
#>  TVKA 0.5   | WT   70

mod3%>%
  ev(amt=75)%>%
  idata_set(idata)%>%
  obsonly()%>%
  #zero_re()%>%
  carry_out(dose)%>%
  mrgsim(start=0, end=24, delta=0.033)%>%
  plot(DV~time)
```

![](https://i.imgur.com/Wo8aNmK.png)

<sup>Created on 2023-10-04 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>
